integrations/jta/jdbc/src/main/java/io/helidon/integrations/jta/jdbc/LocalXAResource.java

```diff
@@ -0,0 +1,1031 @@
+        // https://github.com/jbosstm/narayana/blob/c5f02d07edb34964b64341974ab689ea44536603/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L227-L261.
+        Object association =
+            this.computeAssociation(XARoutine.PREPARE,
+                                    xid,
+                                    EnumSet.of(Association.BranchState.IDLE),
+                                    LocalXAResource::prepare,
+                                    false); // don't remove association on error
+
+        int returnValue = association == null ? XA_RDONLY : XA_OK;
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "prepare", returnValue);
+        }
+        return returnValue;
+    }
+
+    @Override // XAResource
+    public void commit(Xid xid, boolean onePhase) throws XAException {
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.entering(this.getClass().getName(), "commit", new Object[] {xid, onePhase});
+        }
+        requireNonNullXid(xid);
+
+        // Error handling needs to be extraordinarily specific. XAER_RMERR indicates catastrophic failure (like if a
+        // local rollback(), issued in response to a local commit() failure, occurs).  XAER_RMFAIL indicates a transient
+        // error, i.e. we tried to Do The Thing but for now it Didn't Work.
+        //
+        // Concrete examples: You can see that XAER_RMERR and (the completely non-transient) XAER_PROTO (for example)
+        // are both treated as Equally Bad Things:
+        // https://github.com/jbosstm/narayana/blob/c5f02d07edb34964b64341974ab689ea44536603/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L512-L514
+        //
+        // You can also see that XAER_RMFAIL does something different and is no different from XA_RETRY:
+        // https://github.com/jbosstm/narayana/blob/c5f02d07edb34964b64341974ab689ea44536603/ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java#L525-L534
+        //
+        // Finally, if an error happens during commit, we issue a SQL/JDBC/local transaction rollback() command.  If
+        // that works, then we return XA_RB*. If that fails, chances are we return XAER_RMERR, but
+        this.computeAssociation(XARoutine.COMMIT,
+                                xid,
+                                EnumSet.of(Association.BranchState.IDLE,
+                                           Association.BranchState.PREPARED),
+                                a -> commitAndReset(a, onePhase),
+                                true); // do remove association on error
+
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.exiting(this.getClass().getName(), "commit");
+        }
+    }
+
+    @Override // XAResource
+    public void rollback(Xid xid) throws XAException {
```

Comment:
This method is invoked as part of the two-phase completion process (after the `start`/`end` cycle, and possibly after a `prepare` invocation).  Strictly speaking this may be invoked on any thread at any time after the `start`/`end` cycle.  The resource manager must roll its work back.